allennlp/models/encoder_decoders/simple_seq2seq.py

```diff
@@ -27,221 +27,415 @@ class SimpleSeq2Seq(Model):
+            tensors has shape ``(group_size, *)``, where ``*`` can be any other number
+            of dimensions.
+
+        Returns
+        -------
+        Tuple[torch.Tensor, Dict[str, torch.Tensor]]
+            A tuple of ``(log_probabilities, updated_state)``, where ``log_probabilities``
+            is a tensor of shape ``(group_size, num_classes)`` containing the predicted
+            log probability of each class for the next step, for each item in the group,
+            while ``updated_state`` is a dictionary of tensors containing the encoder outputs,
+            source mask, and updated decoder hidden state and context.
+
+        Notes
+        -----
+            We treat the inputs as a batch, even though ``group_size`` is not necessarily
+            equal to ``batch_size``, since the group may contain multiple states
+            for each source sentence in the batch.
+        """
+        # shape: (group_size, num_classes)
+        output_projections, state = self._prepare_output_projections(last_predictions, state)
+
+        # shape: (group_size, num_classes)
+        class_log_probabilities = F.log_softmax(output_projections, dim=-1)
+
+        return class_log_probabilities, state
+
     @overrides
     def forward(self,  # type: ignore
                 source_tokens: Dict[str, torch.LongTensor],
                 target_tokens: Dict[str, torch.LongTensor] = None) -> Dict[str, torch.Tensor]:
         # pylint: disable=arguments-differ
         """
-        Decoder logic for producing the entire target sequence.
+        Make foward pass with decoder logic for producing the entire target sequence.
 
         Parameters
         ----------
-        source_tokens : Dict[str, torch.LongTensor]
-           The output of ``TextField.as_array()`` applied on the source ``TextField``. This will be
-           passed through a ``TextFieldEmbedder`` and then through an encoder.
-        target_tokens : Dict[str, torch.LongTensor], optional (default = None)
-           Output of ``Textfield.as_array()`` applied on target ``TextField``. We assume that the
-           target tokens are also represented as a ``TextField``.
+        source_tokens : ``Dict[str, torch.LongTensor]``
+           The output of `TextField.as_array()` applied on the source `TextField`. This will be
+           passed through a `TextFieldEmbedder` and then through an encoder.
+        target_tokens : ``Dict[str, torch.LongTensor]``, optional (default = None)
+           Output of `Textfield.as_array()` applied on target `TextField`. We assume that the
+           target tokens are also represented as a `TextField`.
+
+        Returns
+        -------
+        Dict[str, torch.Tensor]
+        """
+        state = self._init_encoded_state(source_tokens)
+
+        if target_tokens:
+            return self._forward_loop(target_tokens, state)
+
+        return self._forward_beam_search(state)
```

Comment:
I think running beam search should be guarded behind a flag. Potentially just run it when `beam_size` is defined. The rationale is again backwards compatibility as the shape of the tensors in the returned dictionary is not the same with beam search.